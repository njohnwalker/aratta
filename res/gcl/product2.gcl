@req()
m1, res := m, 0 ;
do @inv()
// invariant m1 >= 0 & res == (m - m1) * n
[] m1 > 0 ->
       n1 := n ;
       do @inv()
       // invariant n1 >= 0 & res == (m - m1 + 1) * n - n1
       [] n1 > 0 -> res, n1 := res + 1, n1 -1 ;
       od
       m1 := m1 - 1 ;
od
@ens(res == m * n)