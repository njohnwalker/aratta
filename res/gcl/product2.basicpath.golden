Path 1
  ; Substitute
      @m1 -> m
      @res -> 0
  { Satisfies INVARIANT == PLACEHOLDER }
Path 2
  ; Assume INVARIANT == PLACEHOLDER
  [ Branch 1
      ; Assume m1 > 0
      ; Substitute
          @n1 -> n
      { Satisfies INVARIANT == PLACEHOLDER }
  ]
  { Satisfies INVARIANT == PLACEHOLDER }
Path 3
  ; Assume INVARIANT == PLACEHOLDER
  [ Branch 1
      ; Assume INVARIANT == PLACEHOLDER
      [ Branch 1
          ; Assume n1 > 0
          ; Substitute
              @res -> res + 1
              @n1 -> n1 - 1
          { HOLE }
      ]
      { Satisfies INVARIANT == PLACEHOLDER }
  ]
  { Satisfies INVARIANT == PLACEHOLDER }
Path 4
  ; Assume INVARIANT == PLACEHOLDER
  [ Branch 1
      ; Assume INVARIANT == PLACEHOLDER
      ; Assume ~ n1 > 0
      ; Substitute
          @m1 -> m1 - 1
      { HOLE }
  ]
  { Satisfies INVARIANT == PLACEHOLDER }
Path 5
  ; Assume INVARIANT == PLACEHOLDER
  ; Assume ~ m1 > 0
  { Satisfies res == m * n }